README.md

```diff
@@ -66,55 +66,65 @@ flowchart LR
 built to make sure there is no gap for the Garbage Collector to exploit. From there, since underlying values of the
-`wafObject` are either arrays (for maps, structs and arrays) or string (for all ints, booleans and strings),
+`wafObject` are either arrays of wafObjects (for maps, structs and arrays) or string (for all ints, booleans and strings),
 we can store 2 slices of arrays and use `runtime.KeepAlive` in each code path to protect them from the GC.
 
+All these objects stored in the reference pool need to live throughout the use of the Waf Context associated.
+
 ### Typical call to Run()
 
 Here is an example of the flow of operations on a simple call to Run():
 
-- Encode input data into Waf Objects
+- Encode input data into Waf Objects and store references in the temporary pool
 - Lock the context mutex until the end of the call
+- store references from the temporary pool into the context level pool
 - Call `ddwaf_run`
 - Decode the matches and actions
 
 ### CGO-less C Bindings
 
-The main component used to build C bindings without using CGO is called [purego](https://github.com/ebitengine/purego). The flow of execution on our side is to embed the C shared library using `go:embed`. Then to dump it into a file, load it using `dlopen` and to load the symbols using `dlsym`. And finally to call them.
+The main component used to build C bindings without using CGO is called [purego](https://github.com/ebitengine/purego). The flow of execution on our side
+is to embed the C shared library using `go:embed`. Then to dump it into a file, load it using `dlopen` and to load the
+symbols using `dlsym`. And finally to call them.
 
 ⚠️ Keep in mind that **purego only works on linux/darwin for amd64/arm64 and so does go-libddwaf.**
 
-Another requirement of `libddwaf` is to have a FHS filesystem on your machine and, for linux, to provide `libc.so.6`, `libpthread.so.0` and `libm.so.6`, `libdl.so.2` as dynamic libraries.
+Another requirement of `libddwaf` is to have a FHS filesystem on your machine and, for linux, to provide `libc.so.6`,
+`libpthread.so.0` and `libm.so.6`, `libdl.so.2` as dynamic libraries.
 
-## Contributing usual pitfalls
+## Contributing pitfalls
 
-- Cannot dlopen twice in the app lifetime on OSX
-- `runtime.KeepAlive()` calls are here to prevent the GC from destroying objects too early
-- Since there is a stack switch between the go code and the C code, usually the only C stacktrace you will ever get is from gdb
-- If a segfault happens during a call to the C code, the goroutine stacktrace which has done the call is the one annotated with `[syscall]`.
+- Cannot dlopen twice in the app lifetime on OSX. It messes with Thread Local Storage and usually finishes with a `std::bad_alloc()`
+- `keepAlive()` calls are here to prevent the GC from destroying objects too early
+- Since there is a stack switch between the go code and the C code, usually the only C stacktrace you will ever get is from GDB
```

Comment:
What does this mean concretely? Means the C code will always be invisible unless you have a gdb attached, and C frames won't unwind all the way into go? TBH this makes me think `CONTRIBUTING.md` should give information on how to debug your code and all...